[Datasets] `test_streaming_split_e2e` is failing

### What happened + What you expected to happen

`test_streaming_integration` has been broken since https://github.com/ray-project/ray/commit/694a74acedb17fc3a07fb0972404a46f7660deb1. Tests passed on the PR but are failing on master, so this is possibly merge-related.

```
ray_start_10_cpus_shared = RayContext(dashboard_url='127.0.0.1:8265', python_version='3.7.9', ray_version='3.0.0.dev0', ray_commit='{{RAY_COMMIT_...3:48107', 'dashboard_agent_listen_port': 52365, 'node_id': '5bbb1fc4077cfb176948562247f805555fd503a170ab8b59dc893014'})

    def test_streaming_split_e2e(ray_start_10_cpus_shared):
        def get_lengths(*iterators):
            lengths = []
            for it in iterators:
                x = 0
                for batch in it.iter_batches():
                    x += len(batch)
                lengths.append(x)
            lengths.sort()
            return lengths
    
        ds = ray.data.range(1000)
        (
            i1,
            i2,
        ) = ds.streaming_split(2, equal=True)
>       lengths = get_lengths(i1, i2)

python/ray/data/tests/test_streaming_integration.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
python/ray/data/tests/test_streaming_integration.py:96: in get_lengths
    for batch in it.iter_batches():
/ray/python/ray/data/_internal/stream_split_dataset_iterator.py:115: in iter_batches
    shuffle_seed=local_shuffle_seed,
/ray/python/ray/data/_internal/block_batching.py:123: in batch_block_refs
    prefetch_batches=prefetch_batches,
/ray/python/ray/data/_internal/block_batching.py:172: in batch_blocks
    for formatted_batch in batch_iter:
/ray/python/ray/data/_internal/block_batching.py:398: in _format_batches
    for block in block_iter:
/ray/python/ray/data/_internal/block_batching.py:361: in _blocks_to_batches
    for block in block_iter:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

block_ref_iter = <generator object _prefetch_blocks at 0x7f9b303c2750>
stats = None

    def _resolve_blocks(
        block_ref_iter: Iterator[ObjectRef[Block]],
        stats: Optional[Union[DatasetStats, DatasetPipelineStats]] = None,
    ) -> Iterator[Block]:
        """Given an iterator of unresolved blocks (as Ray object references), returns an
        iterator of resolved blocks.
    
        The length of the returned iterator may be less than the length of the original
        if any of the references in the original iterator are None.
    
        Args:
            block_ref_iter: An iterator over block object references.
            stats: Dataset stats object used to store block fetching time.
    
        Returns:
            An iterator over resolved blocks.
        """
    
        hit = 0
        miss = 0
        unknown = 0
        for block_ref in block_ref_iter:
            if block_ref is not None:
                stats_timer = stats.iter_get_s.timer() if stats else nullcontext()
                # Count the number of blocks that we hit locally or miss (so have to
                # fetch from remote node). This is to measure the effectiveness of
                # prefetch.
                loc = ray.experimental.get_object_locations([block_ref])
                nodes = loc[block_ref]["node_ids"]
                if nodes:
                    current = ray.get_runtime_context().get_node_id()
                    if current in nodes:
                        hit += 1
                    else:
                        miss += 1
                else:
                    unknown += 1
                with stats_timer:
                    block = ray.get(block_ref)
                yield block
>       stats.iter_blocks_local = hit
E       AttributeError: 'NoneType' object has no attribute 'iter_blocks_local'

/ray/python/ray/data/_internal/block_batching.py:265: AttributeError
```

### Versions / Dependencies

Master

### Reproduction script

See above.

### Issue Severity

None

--- a/python/ray/data/_internal/dataset_iterator_impl.py
+++ b/python/ray/data/_internal/dataset_iterator_impl.py
@@ -2,9 +2,8 @@ from typing import TYPE_CHECKING, Optional, Union, Iterator, Callable, Any
 import time
 import warnings
 
-from ray.air.util.data_batch_conversion import BlockFormat
+from ray.data._internal.util import _get_batch_format
 from ray.data.block import BlockAccessor, T
-from ray.data.context import DatasetContext
 from ray.data.row import TableRow
 from ray.data.block import DataBatch
 from ray.data.dataset_iterator import DatasetIterator
@@ -60,24 +59,7 @@ class DatasetIteratorImpl(DatasetIterator):
         # During row-based ops, we also choose a batch format that lines up with the
         # current dataset format in order to eliminate unnecessary copies and type
         # conversions.
-        ctx = DatasetContext.get_current()
-        if ctx.use_streaming_executor:
-            # TODO: calling dataset_format() triggers bulk execution.
-            batch_format = "default"
-        else:
-            try:
-                dataset_format = self._base_dataset.dataset_format()
-            except ValueError:
-                # Dataset is empty or cleared, so fall back to "default".
-                batch_format = "default"
-            else:
-                batch_format = (
-                    "pyarrow"
-                    if dataset_format == BlockFormat.ARROW
-                    else "pandas"
-                    if dataset_format == BlockFormat.PANDAS
-                    else "default"
-                )
+        batch_format = _get_batch_format(self._base_dataset)
         for batch in self.iter_batches(
             batch_size=None, prefetch_blocks=prefetch_blocks, batch_format=batch_format
         ):

--- a/python/ray/data/_internal/dataset_iterator_impl.py
+++ b/python/ray/data/_internal/dataset_iterator_impl.py
@@ -3,7 +3,7 @@ import time
 import warnings
 import sys
 
-from ray.data._internal.util import _best_batch_format
+from ray.data._internal.util import _default_batch_format
 from ray.data.block import DataBatch
 from ray.data.dataset_iterator import DatasetIterator
 from ray.data._internal.block_batching import batch_block_refs
@@ -65,8 +65,8 @@ class DatasetIteratorImpl(DatasetIterator):
     def schema(self) -> Union[type, "pyarrow.lib.Schema"]:
         return self._base_dataset.schema()
 
-    def _best_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
-        return _best_batch_format(self._base_dataset)
+    def _default_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
+        return _default_batch_format(self._base_dataset)
 
     def __getattr__(self, name):
         if name == "_base_dataset":

--- a/python/ray/data/_internal/dataset_iterator_impl.py
+++ b/python/ray/data/_internal/dataset_iterator_impl.py
@@ -1,10 +1,9 @@
 from typing import TYPE_CHECKING, Optional, Union, Iterator, Callable, Any
 import time
 import warnings
+import sys
 
-from ray.data._internal.util import _get_batch_format
-from ray.data.block import BlockAccessor, T
-from ray.data.row import TableRow
+from ray.data._internal.util import _best_batch_format
 from ray.data.block import DataBatch
 from ray.data.dataset_iterator import DatasetIterator
 from ray.data._internal.block_batching import batch_block_refs
@@ -12,6 +11,11 @@ from ray.data._internal.block_batching import batch_block_refs
 if TYPE_CHECKING:
     import pyarrow
     from ray.data import Dataset
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
 
 
 class DatasetIteratorImpl(DatasetIterator):
@@ -55,23 +59,14 @@ class DatasetIteratorImpl(DatasetIterator):
 
         stats.iter_total_s.add(time.perf_counter() - time_start)
 
-    def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterator[Union[T, TableRow]]:
-        # During row-based ops, we also choose a batch format that lines up with the
-        # current dataset format in order to eliminate unnecessary copies and type
-        # conversions.
-        batch_format = _get_batch_format(self._base_dataset)
-        for batch in self.iter_batches(
-            batch_size=None, prefetch_blocks=prefetch_blocks, batch_format=batch_format
-        ):
-            batch = BlockAccessor.for_block(BlockAccessor.batch_to_block(batch))
-            for row in batch.iter_rows():
-                yield row
-
     def stats(self) -> str:
         return self._base_dataset.stats()
 
     def schema(self) -> Union[type, "pyarrow.lib.Schema"]:
         return self._base_dataset.schema()
+
+    def _best_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
+        return _best_batch_format(self._base_dataset)
 
     def __getattr__(self, name):
         if name == "_base_dataset":

--- a/python/ray/data/_internal/pipelined_dataset_iterator.py
+++ b/python/ray/data/_internal/pipelined_dataset_iterator.py
@@ -1,8 +1,7 @@
 from typing import TYPE_CHECKING, Any, Callable, Optional, Union, Iterator
 import warnings
 
-from ray.data.block import DataBatch, T
-from ray.data.row import TableRow
+from ray.data.block import DataBatch
 from ray.data.dataset_iterator import DatasetIterator
 
 if TYPE_CHECKING:
@@ -48,12 +47,6 @@ class PipelinedDatasetIterator(DatasetIterator):
             local_shuffle_buffer_size=local_shuffle_buffer_size,
             local_shuffle_seed=local_shuffle_seed,
             _collate_fn=_collate_fn,
-        )
-
-    def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterator[Union[T, TableRow]]:
-        ds = self._get_next_dataset()
-        return ds.iter_rows(
-            prefetch_blocks=prefetch_blocks,
         )
 
     def stats(self) -> str:

--- a/python/ray/data/_internal/stream_split_dataset_iterator.py
+++ b/python/ray/data/_internal/stream_split_dataset_iterator.py
@@ -14,10 +14,12 @@ from typing import (
 )
 
 import ray
+from ray.data._internal.util import _get_batch_format
 
 from ray.data.dataset_iterator import DatasetIterator
-from ray.data.block import Block, DataBatch
+from ray.data.block import Block, BlockAccessor, DataBatch, T
 from ray.data.context import DatasetContext
+from ray.data.row import TableRow
 from ray.data._internal.execution.streaming_executor import StreamingExecutor
 from ray.data._internal.execution.legacy_compat import (
     execute_to_legacy_bundle_iterator,
@@ -114,6 +116,18 @@ class StreamSplitDatasetIterator(DatasetIterator):
             shuffle_buffer_min_size=local_shuffle_buffer_size,
             shuffle_seed=local_shuffle_seed,
         )
+
+    def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterator[Union[T, TableRow]]:
+        # During row-based ops, we also choose a batch format that lines up with the
+        # current dataset format in order to eliminate unnecessary copies and type
+        # conversions.
+        batch_format = _get_batch_format(self._base_dataset)
+        for batch in self.iter_batches(
+            batch_size=None, prefetch_blocks=prefetch_blocks, batch_format=batch_format
+        ):
+            batch = BlockAccessor.for_block(BlockAccessor.batch_to_block(batch))
+            for row in batch.iter_rows():
+                yield row
 
     def stats(self) -> str:
         """Implements DatasetIterator."""

--- a/python/ray/data/_internal/stream_split_dataset_iterator.py
+++ b/python/ray/data/_internal/stream_split_dataset_iterator.py
@@ -14,12 +14,11 @@ from typing import (
 )
 
 import ray
-from ray.data._internal.util import _get_batch_format
+from ray.data._internal.util import _best_batch_format
 
 from ray.data.dataset_iterator import DatasetIterator
-from ray.data.block import Block, BlockAccessor, DataBatch, T
+from ray.data.block import Block, DataBatch
 from ray.data.context import DatasetContext
-from ray.data.row import TableRow
 from ray.data._internal.execution.streaming_executor import StreamingExecutor
 from ray.data._internal.execution.legacy_compat import (
     execute_to_legacy_bundle_iterator,
@@ -117,18 +116,6 @@ class StreamSplitDatasetIterator(DatasetIterator):
             shuffle_seed=local_shuffle_seed,
         )
 
-    def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterator[Union[T, TableRow]]:
-        # During row-based ops, we also choose a batch format that lines up with the
-        # current dataset format in order to eliminate unnecessary copies and type
-        # conversions.
-        batch_format = _get_batch_format(self._base_dataset)
-        for batch in self.iter_batches(
-            batch_size=None, prefetch_blocks=prefetch_blocks, batch_format=batch_format
-        ):
-            batch = BlockAccessor.for_block(BlockAccessor.batch_to_block(batch))
-            for row in batch.iter_rows():
-                yield row
-
     def stats(self) -> str:
         """Implements DatasetIterator."""
         return self._base_dataset.stats()
@@ -136,6 +123,9 @@ class StreamSplitDatasetIterator(DatasetIterator):
     def schema(self) -> Union[type, "pyarrow.lib.Schema"]:
         """Implements DatasetIterator."""
         return self._base_dataset.schema()
+
+    def _best_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
+        return _best_batch_format(self._base_dataset)
 
 
 @ray.remote(num_cpus=0)

--- a/python/ray/data/_internal/stream_split_dataset_iterator.py
+++ b/python/ray/data/_internal/stream_split_dataset_iterator.py
@@ -14,7 +14,7 @@ from typing import (
 )
 
 import ray
-from ray.data._internal.util import _best_batch_format
+from ray.data._internal.util import _default_batch_format
 
 from ray.data.dataset_iterator import DatasetIterator
 from ray.data.block import Block, DataBatch
@@ -124,8 +124,8 @@ class StreamSplitDatasetIterator(DatasetIterator):
         """Implements DatasetIterator."""
         return self._base_dataset.schema()
 
-    def _best_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
-        return _best_batch_format(self._base_dataset)
+    def _default_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
+        return _default_batch_format(self._base_dataset)
 
 
 @ray.remote(num_cpus=0)

--- a/python/ray/data/_internal/util.py
+++ b/python/ray/data/_internal/util.py
@@ -9,10 +9,17 @@ import numpy as np
 
 import ray
 from ray.air.constants import TENSOR_COLUMN_NAME
+from ray.air.util.data_batch_conversion import BlockFormat
 from ray.data.context import DatasetContext
 from ray._private.utils import _get_pyarrow_version
 
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
 if TYPE_CHECKING:
+    from ray.data.dataset import Dataset
     from ray.data.datasource import Reader
     from ray.util.placement_group import PlacementGroup
 
@@ -397,3 +404,28 @@ def _split_list(arr: List[Any], num_splits: int) -> List[List[Any]]:
         arr[i * q + min(i, r) : (i + 1) * q + min(i + 1, r)] for i in range(num_splits)
     ]
     return splits
+
+
+def _get_batch_format(
+    ds: "Dataset",
+) -> Literal["default", "pandas", "pyarrow", "numpy"]:
+    """Get a batch format that lines up with the dataset format."""
+    ctx = DatasetContext.get_current()
+    if ctx.use_streaming_executor:
+        # TODO: calling dataset_format() triggers bulk execution.
+        batch_format = "default"
+    else:
+        try:
+            dataset_format = ds.dataset_format()
+        except ValueError:
+            # Dataset is empty or cleared, so fall back to "default".
+            batch_format = "default"
+        else:
+            batch_format = (
+                "pyarrow"
+                if dataset_format == BlockFormat.ARROW
+                else "pandas"
+                if dataset_format == BlockFormat.PANDAS
+                else "default"
+            )
+    return batch_format

--- a/python/ray/data/_internal/util.py
+++ b/python/ray/data/_internal/util.py
@@ -406,7 +406,7 @@ def _split_list(arr: List[Any], num_splits: int) -> List[List[Any]]:
     return splits
 
 
-def _get_batch_format(
+def _best_batch_format(
     ds: "Dataset",
 ) -> Literal["default", "pandas", "pyarrow", "numpy"]:
     """Get a batch format that lines up with the dataset format."""

--- a/python/ray/data/_internal/util.py
+++ b/python/ray/data/_internal/util.py
@@ -406,10 +406,10 @@ def _split_list(arr: List[Any], num_splits: int) -> List[List[Any]]:
     return splits
 
 
-def _best_batch_format(
+def _default_batch_format(
     ds: "Dataset",
 ) -> Literal["default", "pandas", "pyarrow", "numpy"]:
-    """Get a batch format that lines up with the dataset format."""
+    """Get the best batch format that lines up with the dataset format."""
     ctx = DatasetContext.get_current()
     if ctx.use_streaming_executor:
         # TODO: calling dataset_format() triggers bulk execution.

--- a/python/ray/data/dataset_iterator.py
+++ b/python/ray/data/dataset_iterator.py
@@ -126,7 +126,7 @@ class DatasetIterator(abc.ABC):
         Returns:
             An iterator over rows of the dataset.
         """
-        target_format = self._best_batch_format()
+        target_format = self._default_batch_format()
         for batch in self.iter_batches(
             batch_size=None, prefetch_blocks=prefetch_blocks, batch_format=target_format
         ):
@@ -705,7 +705,7 @@ class DatasetIterator(abc.ABC):
             return False
         return _is_tensor_schema(schema.names)
 
-    def _best_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
+    def _default_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
         """Returns the best batch format that lines up with the dataset format.
 
         NOTE: Return "default" here. Subclass can override this method to decide best

--- a/python/ray/data/dataset_iterator.py
+++ b/python/ray/data/dataset_iterator.py
@@ -3,7 +3,7 @@ import numpy as np
 import sys
 from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Union, Iterator
 
-from ray.data.block import DataBatch, T
+from ray.data.block import BlockAccessor, DataBatch, T
 from ray.data.row import TableRow
 from ray.util.annotations import PublicAPI
 from ray.data._internal.util import _is_tensor_schema
@@ -104,7 +104,6 @@ class DatasetIterator(abc.ABC):
         """
         raise NotImplementedError
 
-    @abc.abstractmethod
     def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterator[Union[T, TableRow]]:
         """Return a local row iterator over the dataset.
 
@@ -127,8 +126,13 @@ class DatasetIterator(abc.ABC):
         Returns:
             An iterator over rows of the dataset.
         """
-
-        raise NotImplementedError
+        target_format = self._best_batch_format()
+        for batch in self.iter_batches(
+            batch_size=None, prefetch_blocks=prefetch_blocks, batch_format=target_format
+        ):
+            batch = BlockAccessor.for_block(BlockAccessor.batch_to_block(batch))
+            for row in batch.iter_rows():
+                yield row
 
     @abc.abstractmethod
     def stats(self) -> str:
@@ -700,3 +704,11 @@ class DatasetIterator(abc.ABC):
         if schema is None or isinstance(schema, type):
             return False
         return _is_tensor_schema(schema.names)
+
+    def _best_batch_format(self) -> Literal["default", "pandas", "pyarrow", "numpy"]:
+        """Returns the best batch format that lines up with the dataset format.
+
+        NOTE: Return "default" here. Subclass can override this method to decide best
+        batch format.
+        """
+        return "default"

--- a/python/ray/data/tests/test_streaming_integration.py
+++ b/python/ray/data/tests/test_streaming_integration.py
@@ -1,3 +1,4 @@
+import itertools
 import pytest
 import threading
 import time
@@ -89,12 +90,16 @@ def test_output_split_e2e(ray_start_10_cpus_shared):
 
 
 def test_streaming_split_e2e(ray_start_10_cpus_shared):
-    def get_lengths(*iterators):
+    def get_lengths(*iterators, use_iter_batches=True):
         lengths = []
         for it in iterators:
             x = 0
-            for batch in it.iter_batches():
-                x += len(batch)
+            if use_iter_batches:
+                for batch in it.iter_batches():
+                    x += len(batch)
+            else:
+                for _ in it.iter_rows():
+                    x += 1
             lengths.append(x)
         lengths.sort()
         return lengths
@@ -124,13 +129,15 @@ def test_streaming_split_e2e(ray_start_10_cpus_shared):
     assert lengths == [0, 1], lengths
 
     ds = ray.data.range(1000, parallelism=10)
-    i1, i2, i3 = ds.streaming_split(3, equal=True)
-    lengths = get_lengths(i1, i2, i3)
-    assert lengths == [333, 333, 333], lengths
-
-    i1, i2, i3 = ds.streaming_split(3, equal=False)
-    lengths = get_lengths(i1, i2, i3)
-    assert lengths == [300, 300, 400], lengths
+    for equal_split, use_iter_batches in itertools.product(
+        [True, False], [True, False]
+    ):
+        i1, i2, i3 = ds.streaming_split(3, equal=equal_split)
+        lengths = get_lengths(i1, i2, i3, use_iter_batches=use_iter_batches)
+        if equal_split:
+            assert lengths == [333, 333, 333], lengths
+        else:
+            assert lengths == [300, 300, 400], lengths
 
 
 def test_e2e_option_propagation(ray_start_10_cpus_shared, restore_dataset_context):

